### ISSUE TITLE
Decide the environment where !! is evaluated

### ISSUE THREAD

**Issue by wlandau** [tidyverse/rlang#399]
I am trying to make [the `drake` package](https://github.com/ropensci/drake) play nicely with tidy evaluation (ropensci/drake#200), but I seem to be having scoping issues. Here is a non-[`drake`](https://github.com/ropensci/drake) minimal working example.

```r
envir <- new.env(parent = globalenv())
eval(parse(text = "little_b <- 'b'"), envir = envir)
eval(rlang::expr(!!little_b), envir = envir) # Desired result: "b"

## Error in (function (x)  : object 'little_b' not found
```

How do I take control of the environment where `rlang` looks for symbols? I realize that `rlang` has its own non-lexical scoping mechanism, and I apologize for not entirely grasping it.

**Comment by egnha** [comment]
Since you want to evaluate the expression `rlang::expr(!!little_b)` literally in `envir` (where `little_b` is visible), you need to quote it before passing it to `eval()`. Have you tried using `evalq()` in place of `eval()`?

**Comment by lionel-** [comment]
First of all let's replace the parse with something more idiomatic ;)

```r
envir <- new.env(parent = globalenv())
envir$little_b <- "b"
```

`!!` exclusively evaluates in the lexical scope. And it evaluates immediately, no need for `eval()`:

```r
expr(!!little_b)  # Error because `little_b` is not defined in this context
```

You _might_ want to quote your quasiquoting expression and then evaluate this in a different context:

```r
x <- quote(rlang::expr(!!little_b))
eval(x, envir)
```

Like Eugene suggests `evalq()` would also work for this:

```r
evalq(quote(rlang::expr(!!little_b)), envir)
```

But I think this suggests you're going about it the wrong way. Quoting a quasiquoted expression is complicated to reason about and almost never needed.

**Comment by wlandau** [comment]
Thanks both for your help. I think I solved the issue in ropensci/drake@ccd967c9f08a4400eea59f40219325871fde7088, and it did turn out to be kind of like `evalq(rlang::expr(!!little_b), envir)`. I am not actually using `evalq()` because I rely on `evaluate::try_capture_stack()`. For the general case, the code actually turns out to look something like

```r
evaluate::try_capture_stack(quote(rlang::expr(!!little_b)), envir)
```

The reason is that I need the complete error log if the code inside `quote()` fails.
